### ISSUE TITLE
Remote server URLs don't work for `pdfDefaultOptions.assetsFolder`

### ISSUE THREAD

**Issue by etramell** [stephanrauh/ngx-extended-pdf-viewer#1478]
**Check this first**

Sorry, but the nature of this problem means my answers to your three questions are:
- Probably not.
- Probably not.
- Um, I don't know how for this one...

**Background info before I can describe the bug...**

We are using your excellent PDF viewer in a multi-tenant Angular app (one app white-labelled and served separately for each of many customers).  Without too much detail that's irrelevant to you, just know that our architecture causes our `assets` path to be not relative to the site URL, but on a completely separate server.

To illustrate this, here are a couple of pretend URLs for a couple of sites using our app:
- `https://customer1.our-app.com`
- `https://customer2.our-app.com`

And a pretend assets URL shared by all sites:
- `https://library.our-app.com/assets`

When we upgraded to your v13 (along with our upgrade to Angular v13), we could no longer load PDFs:
<img width="769" alt="image" src="https://user-images.githubusercontent.com/1588343/182980212-a12df6fe-685d-497c-98de-85cb5a5c6f81.png">

We determined that this was due to the fact that your code was now trying to load the new `op-chaining-support` from a relative `assets` URL, and it couldn't, since those aren't there.  (The browser was trying to parse our app's error page as that JavaScript file, because of Angular default routing.  THAT was confusing!)

Further research through your GitHub issue discussions taught us that we just needed to set `pdfDefaultOptions.assetsFolder`, so we updated our code so that it basically has this outcome:
```ts
pdfDefaultOptions.assetsFolder = 'https://library.our-app.com/assets';
```
This fixed the `op-chaining-support` problem.

However...

**Describe the bug**

We then got this message when trying to load:
<img width="671" alt="image" src="https://user-images.githubusercontent.com/1588343/182975876-d33733e0-8408-4b4d-b239-8627bc2cbd3f.png">

I determined that we were trying to load the PDF worker via a URL like `https://customer1.our-app.com/https://library.our-app.com/assets/pdf.worker-2.13.491.min.js`.  (Note the double `https`.)

I found this in your source code in `pdf-default-options.ts`:
```ts
  workerSrc: () =>
    pdfDefaultOptions.needsES5
      ? `./${pdfDefaultOptions.assetsFolder}/pdf.worker-${getVersionSuffix(pdfDefaultOptions.assetsFolder)}-es5.js`
      : `./${pdfDefaultOptions.assetsFolder}/pdf.worker-${getVersionSuffix(pdfDefaultOptions.assetsFolder)}.js`,
```
(Ditto for other options -- `cMapUrl`, `sandboxBundleSrc`, `standardFontDataUrl`.)

You've made the very reasonable assumption that assets are always served via a relative URL, but this doesn't work for our particular use case.

**Workaround**

You've given us the ability to override these other options, so that's what I'm doing.  I don't want to recalculate everything, though, so I'm doing (something like) this:
```ts
pdfDefaultOptions.assetsFolder = 'https://library.our-app.com/assets';

const newWorkerSrcReturnValue: string = pdfDefaultOptions.workerSrc().replace(/^\.\//, '');
pdfDefaultOptions.workerSrc = () => newWorkerSrcReturnValue;

// Ditto for the other options that depend on pdfDefaultOptions.assetsFolder.
```
And this seems to work.

**Proposed solution**

Make your default options functions account for a possible full remote URL in `pdfDefaultOptions.assetsFolder`, and leave off the `./` in that case?

But that's up to you...  Not sure if our use case is truly a one-off or not.

If you decide to just call this a "won't fix", I don't blame you.  But at least this is documented in your issues, in case anyone else out there happens to have a similar setup.  And our workaround should be fine for us (and future them, perhaps).

But if you do decide to support this weird use case, we'll be able to delete a bit of workaround code in the future.  🙂 

**And a question:**

Do you know of any other complications that my workaround won't prevent?

**Version info**
 - v13.0.0

**Desktop (please complete the following information):**
 - Probably irrelevant, but mostly Chrome
 
**Smartphone (please complete the following information):**
 - Nothing specific

**To Reproduce**
  - Can you provide a reproducer?
    - Sorry, I don't think so.

  - If you really can't provide me a reproducer, please tell me how to reproduce the behavior:
    - Hopefully the description above is sufficient.

**Thank you!**
  - No... thank YOU!  You've got a great PDF viewer here -- and Angular ready!  😍

**Comment by stephanrauh** [comment]
@Snuffleupagus @calixteman @timvandermeij I'm receiving a lot of positive feedback for wrapping your work into a decent Angular library. It's time to pass the kudos to you, too. Thanks a lot for creating such a great Firefox PDF viewer and even allowing other people to build upon your work! You guys rock!

(And before you ask: please don't try to answer this particular GitHub issue - it's specific to my Angular wrapper. I just wanted to share some of the "thank you"s I get with you).

**Comment by stephanrauh** [comment]
@etramell-tem To be honest, I didn't spend much thought on your scenario yet because it almost always results in a CORS issues, and most people have a hard time solving it. Plus, loading resources from a CDN is more or less illegal if you're in Europe (because of the GDPR), so I tend to discourage it.

Nonetheless, I agree it's annoying it doesn't work. Your description sounds like a corporate CDN, and there's nothing wrong with that.

I'm pretty sure your workaround works well. At the moment, I'm not sure if I can solve the root cause of the problem, or if I prefer to simply document it. Relative paths in Angular applications are always a pain in the - well, you know where. There's always a corner case I didn't think of!

**Comment by stephanrauh** [comment]
Version 15.0.0-alpha.1 fixes the bug. Please run a test to confirm this.

BTW, thanks a lot for your detailed error description!

Best regards,
Stephan

**Comment by stephanrauh** [comment]
Vacations are important, too! :) You might be interested to learn there's still something wrong with the paths. But if I'm not mistaken, it's been wrong for a very long time, and it "only" affects East-Asian fonts. ("Only" meaning I considering the bug very annoying and embarrassing, but the majority of PDF files - even including languages like Japanese, Korean, and Chinese - render without problems). Here's the pointer to the ticket: #1485

**Comment by stephanrauh** [comment]
Thanks a lot, Ed! As for the release planning: at the moment, I'm both a bit lazy and busy doing other stuff, so version 15.0.0 probably get a few weeks to get mature. A major version update is always a good opportunity to add new features, so maybe I'll add an option to enable the brand-new (und unfinished) PDF editor. I haven't seen it myself yet, apart from a very early stage, so I won't promise anything. But if I unlock it, I'm going to play around with the editor for a couple of evenings.

I'm definitely going to support Angular 15 (unless something unexpected happens), but that version is probably still months away. My best guess is it's going to be release in November - but don't take my work for it, it's just an educated guess. By the time Angular 15 lands, ngx-extended-pdf-viewer is probably at version 18. :) That's the drawback of semantic versioning - every breaking change, no matter how small or insignificant, forces me to increase the major version number, even if I don't add new features!

**Comment by stephanrauh** [comment]
As for the embarrassing bugs: yeah, you're right, there are very few bugs I'm proud of. :) However, I wonder if the bug has something to do with the falling download figures. They've plateaued at 40.000 per week, which is just incredible, but with a slight tendency southwards. The other two Angular PDF viewer show similar trends, so maybe it's a general development. However, me accidentally frustrating developers in the far East is a good explanation, too.

<img width="931" alt="image" src="https://user-images.githubusercontent.com/3045767/184978419-94f809b1-1974-474c-aa82-0f5a2c489d4f.png">

**Comment by Jadamae77** [comment]
I have the same scenario. Thanks @etramell-tem for an incredibly well-documented issue, and @stephanrauh for a phenomenal library!

**Comment by stephanrauh** [comment]
@Jadamae77 First of all, thanks for all the kind words! You've said you've got the same scenario. Does the latest alpha version of my library solve your issue? Or are you still suffering from the bug?

**Comment by Jadamae77** [comment]
Yes, the latest version fixes the issue! I wasn't sure on Friday, but we have the files uploaded to our internal CDN and it all works as expected. Thanks again!

**Comment by stephanrauh** [comment]
Wait, what? Now that's really unexpected. But don't worry about bothering me - quite the contrary, I feel honored. Knowing that people care enough about my library to report bugs, even adding a lot of insight - that's just awesome!

I'm afraid I'll have to postpone work on your bug until the end of October. In the meantime, I'm busy doing other things. If you don't hear from me at the end of this month, please nudge me.

**Comment by stephanrauh** [comment]
Oh no! I'm sorry to hear that! My company is looking for new developers, but I'm afraid you don't live in Germany, and employing people abroad is probably too difficult for us (and for you). However, if you want to give it a try, just ping me. Of course, I can't promise anything beyond giving you the opportunity. But that's better than nothing!

In any case, I'm sure you'll find a new job you'll love soon!

Good luck
Stephan

**Comment by stephanrauh** [comment]
Finally, the bugfix has landed with version 17.4.1.
